password_security/tests/test_change_password.py

```diff
@@ -0,0 +1,136 @@
+        res_post.raise_for_status()
+
+        return res_post
+
+    def test_01_empty_password_fail(self):
+        """It should fail when changing password to empty"""
+
+        # Log in: ensure we end up on the right page
+        res_login = self.login("admin", "admin")
+        self.assertEqual(res_login.request.path_url, "/web")
+        self.assertEqual(res_login.status_code, 200)
+
+        # Change password
+        user = self.env["res.users"].search([("login", "=", "admin")], limit=1)
+        self.assertTrue(user)
+        with self.assertRaises(UserError):
+            # UserError: Setting empty passwords is not allowed for security reasons
+            user._change_password("")
+
+    def test_02_change_password_fail(self):
+        """It should fail when changing password to weak"""
+
+        # Log in: ensure we end up on the right page
+        res_login = self.login("admin", "admin")
+        self.assertEqual(res_login.request.path_url, "/web")
+        self.assertEqual(res_login.status_code, 200)
+
+        # Change password: error is raised because it's too short
+        user = self.env["res.users"].search([("login", "=", "admin")], limit=1)
+        self.assertTrue(user)
+        with self.assertRaises(UserError):
+            user._change_password("admin")
+
+        # Change password: error is raised by check password rules
+        with self.assertRaises(ValidationError):
+            user._change_password("adminadmin")
+
+    def test_03_change_password_session_expired(self):
+        """Session expires when password is changed"""
+
+        # Log in: ensure we end up on the right page
+        res_login = self.login("admin", "admin")
+        self.assertEqual(res_login.request.path_url, "/web")
+        self.assertEqual(res_login.status_code, 200)
+
+        # Reload page: ensure we stay on the same page
+        res_page1 = self.url_open("/web")
+        res_page1.raise_for_status()
+        self.assertEqual(res_page1.request.path_url, "/web")
+        self.assertEqual(res_page1.status_code, 200)
+
+        # Change password: no error raised
+        user = self.env["res.users"].search([("login", "=", "admin")], limit=1)
+        self.assertTrue(user)
+        user._change_password("!asdQWE12345_3")
+
+        # Try to reload page: user kicked out
+        res_page2 = self.url_open("/web")
+        res_page2.raise_for_status()
+        self.assertEqual(res_page2.request.path_url, "/web/login")
+        self.assertEqual(res_page2.status_code, 200)
+
+    def test_04_change_password_check_password_history(self):
+        """It should fail when chosen password was previously used"""
+
+        # Set password history limit
+        user = self.env["res.users"].search([("login", "=", "admin")], limit=1)
+        user.company_id.password_history = 3
+        self.assertEqual(len(user.password_history_ids), 0)
+
+        # Change password: password history records created
+        user._change_password("!asdQWE12345_4")
```

Comment:
Odoo stores datetime fields with time granularity of seconds. The tests are failing because the _change_password is processed fast enough for some of the history to be within the same second.

You need to add either sleeps or manually modify the 'date' field in the password_history_ids to be distinct, otherwise it will be ordered by id asc. You could also add id desc as a sort of fallback to the _order.